brainiak/reprsimil/brsa.py

```diff
@@ -1659,3 +2460,1350 @@ def _loglike_AR1_singpara(self, param, XTX, XTDX, XTFX, YTY_diag,
+        # dimension: space,
+        LL = - np.sum(np.log(sigma2)) * (n_T - n_X0) * 0.5 \
+            + np.sum(np.log(1 - rho1**2)) * n_run * 0.5 \
+            - np.sum(np.log(np.linalg.det(X0TAX0))) * 0.5 \
+            - (n_T - n_X0) * n_V * (1 + np.log(2 * np.pi)) * 0.5
+        # The following are for calculating the derivative to a1
+        deriv_a1 = np.empty(n_V)
+        dYTAY_drho1 = self._make_sandwidge_grad(YTDY_diag, YTFY_diag, rho1)
+        # dimension: space,
+        dX0TAX0_drho1 = - X0TDX0 \
+            + 2 * rho1[:, np.newaxis, np.newaxis] * X0TFX0
+        # dimension: space*rank*rank
+        dX0TAY_drho1 = self._make_sandwidge_grad(X0TDY, X0TFY, rho1)
+        # dimension: rank*space
+
+        # The following are executed for each voxel.
+        for i_v in range(n_V):
+            # All variables with _ele as suffix are for data of just one voxel
+            invX0TAX0_X0TAY_ele = np.dot(X0TAX0_i[i_v, :, :], X0TAY[:, i_v])
+            # preparation for the variable below
+            dYTAcorrY_drho1_ele = dYTAY_drho1[i_v] \
+                - np.dot(dX0TAY_drho1[:, i_v], invX0TAX0_X0TAY_ele) * 2\
+                + np.dot(np.dot(invX0TAX0_X0TAY_ele, dX0TAX0_drho1[i_v, :, :]),
+                         invX0TAX0_X0TAY_ele)
+            deriv_a1[i_v] = 2 / np.pi / (1 + a1[i_v]**2) \
+                * (- n_run * rho1[i_v] / (1 - rho1[i_v]**2)
+                   - np.einsum('ij,ij', X0TAX0_i[i_v, :, :],
+                               dX0TAX0_drho1[i_v, :, :]) * 0.5
+                   - dYTAcorrY_drho1_ele * 0.5 / sigma2[i_v])
+        deriv = deriv_a1
+
+        return -LL, -deriv
+
+
+class GBRSA(BRSA):
+    """Group Bayesian representational Similarity Analysis (GBRSA)
+
+    Given the time series of neural imaging data in a region of interest
+    (ROI) and the hypothetical neural response (design matrix) to
+    each experimental condition of interest,
+    calculate the shared covariance matrix of
+    the voxels(recording unit)' response to each condition,
+    and the relative SNR of each voxels.
+    The relative SNR could be considered as the degree of contribution
+    of each voxel to this shared covariance matrix.
+    A correlation matrix converted from the covariance matrix
+    will be provided as a quantification of neural representational similarity.
+    The differences of this tool from BRSA are:
+    (1) It allows fitting a shared covariance matrix (which can be converted
+    to similarity matrix) across multiple subjects.
+    This is analogous to SRM under funcalign submodule. Because of using
+    multiple subjects, the result is less noisy.
+    (2) In the fitting process, the SNR and noise parameters are marginalized
+    for each voxel. Therefore, this tool should be faster than BRSA
+    when analyzing an ROI of hundreds to thousands voxels. It does not
+    provide a spatial smoothness prior on SNR though.
+    Both tools provide estimation of SNR and noise parameters at the end,
```

Comment:
Thanks for asking. Those by GBRSA are posterior mean instead of MAP. I will add this to the document.